frontend/src/components/student/AllActivities/ExpeditionTask/ExpeditionWrapper/ExpeditionWrapper.js

```diff
@@ -0,0 +1,93 @@
+import { Spinner } from 'react-bootstrap'
+import { EXPEDITION_STATUS } from './ExpeditionWrapperHelpers'
+import { StudentRoutes } from '../../../../../routes/PageRoutes'
+import ExpeditionService from '../../../../../services/expedition.service'
+import QuestionSelectionDoor from '../QuestionSelectionDoor/QuestionSelectionDoor'
+import QuestionAndOptions from '../QuestionAndOptions/QuestionAndOptions'
+import Timer from '../Timer/Timer'
+// wrapped elements should be:
+
+// -- in ANSWER
+// QuestionAndOptions
+// ClosedQuestionPage
+// OpenQuestionPage
+
+// -- in CHOOSE
+// QuestionSelectionDoor
+
+export function ExpeditionWrapper() {
+  const navigate = useNavigate()
+  const location = useLocation()
+  const { activityId, alreadyStarted } = location.state
+
+  const [expeditionState, setExpeditionState] = useState(undefined)
+
+  const goToSummary = () => {
+    navigate(StudentRoutes.GAME_MAP.GRAPH_TASK.SUMMARY, {
+      state: { expeditionId: activityId, remainingTime: expeditionState.timeRemaining }
+    })
+  }
+
+  // we will pass this function to "lower" components so that we can reload info from endpoint
+  // in wrapper on changes
+  // if it breaks, check whether we don't need to pass activityId in here explicitly
+  const reloadState = useCallback(
+    () =>
+      ExpeditionService.getCurrentState(activityId).then((response) => {
+        setExpeditionState(response)
+      }),
+    [activityId]
+  )
+
+  useEffect(() => {
+    if (activityId) {
+      if (alreadyStarted) {
+        // no need to start the expedition again, just do getInfo
+        reloadState()
+      } else {
+        // else start the expedition and do getInfo then
+        ExpeditionService.setExpeditionStart(activityId).then(() => {
+          reloadState()
+        })
+      }
+    }
+  }, [activityId, alreadyStarted, reloadState])
+
+  if (expeditionState === undefined) {
+    return <Spinner />
+  }
+
+  if (expeditionState.remainingTime <= 0) {
+    goToSummary()
+  }
+
+  if (expeditionState.status === EXPEDITION_STATUS.CHOOSE) {
+    // return changed Question Select screen
+
+    if (expeditionState.questions.length === 0) {
+      //navigate to summary if there are no questions left
+      goToSummary()
+    }
+
+    return (
+      <Timer activityId={activityId} timeToSolveMillis={expeditionState.timeRemaining} endAction={goToSummary}>
+        <QuestionSelectionDoor activityId={activityId} questions={expeditionState.questions} reloadInfo={reloadState} />
+      </Timer>
+    )
+  }
+
+  if (expeditionState.status === EXPEDITION_STATUS.ANSWER) {
+    // return changed Question Answer Screen
+    return (
+      <Timer activityId={activityId} timeToSolveMillis={expeditionState.timeRemaining} endAction={goToSummary}>
+        <QuestionAndOptions
+          activityId={activityId}
+          question={expeditionState.questionDetails}
+          reloadInfo={reloadState}
+        />
+      </Timer>
+    )
+  }
+}
```

Comment:
Nie rób tego w ten sposób ifami bo wygląda to źle, jest mało czytelne. W dodatku ten return spinner przedzielony losową funkcją. Straszny tu bałagan 😅.
Zrób sobie useMemo, do niego wrzuć `switch case` sprawdzający przypadek. W zależności od przypadku zwracaj odpowiedni komponent, ale bez opakowania w `Timer`. W przypadku case'a `CHOOSE` wywal tego ifa, który robi goToSummary i dawaj go w jakimś useEffect poza rozszerzając ifa o sprawdzenie czy typ jest `CHOOSE`. Na samym dole tego całego wrappera daj normalnie return, tam daj `Timer` i do środka wrzuć tego useMemo. Poniżej przykład:
```jsx
useEffect(() => {
    // ps. expeditionState.questions.length === 0 to jest to samo co !expeditionState.questions.length
    if ((expeditionState.status === EXPEDITION_STATUS.CHOOSE && !expeditionState.questions.length) || expeditionState.remainingTime <= 0){
        goToSummary()
    }
}, [expeditionState.remainingTime, expeditionState.questions.length])

const wrapperContent = useMemo(() => {
    if (expeditionState == null){ // nie jestem pewien czy ten if potrzebny, sprawdź sobie
        return <></>
    }
    switch(expeditionState.status){
        case EXPEDITION_STATUS.ANSWER:
            return 
                <QuestionAndOptions
                    activityId={activityId}
                    question={expeditionState.questionDetails}
                    reloadInfo={reloadState}
                />
        case EXPEDITION_STATUS.CHOOSE:
            return 
                <QuestionSelectionDoor 
                    activityId={activityId} 
                    questions={expeditionState.questions} 
                    reloadInfo={reloadState} 
                />
        default: 
            return <></>
    }
}, []) // odpowiednie depsy!

return expeditionState === undefined ? <Spinner animation={'border'} /> : (
    <Timer activityId={activityId} timeToSolveMillis={expeditionState.timeRemaining} endAction={goToSummary}>
        {wrapperContent}
    </Timer>
)
```